<hr>
<h1>Activity 2: UK ROAD ACCIDENT DATA ANALYTICS </h1>
<hr>
<h3>Analyst : Nerjie Angelo Mecantina</h3>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import f_oneway
import warnings
warnings.filterwarnings('ignore')

<h2>Converting Dataset(s) into Dataframe</h2>

<h3>Requirements</h3>
<ul>
    <li>Name of the Dataframes</li>
    <li>Filepath</li>
</ul>

In [2]:
accidents = pd.read_csv('datasets\\uk_road_accident.csv')

In [3]:
accidents

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accidents.columns

Index(['Index', 'Accident_Severity', 'Accident Date', 'Latitude',
       'Light_Conditions', 'District Area', 'Longitude',
       'Number_of_Casualties', 'Number_of_Vehicles', 'Road_Surface_Conditions',
       'Road_Type', 'Urban_or_Rural_Area', 'Weather_Conditions',
       'Vehicle_Type'],
      dtype='object')

In [5]:
accidents.columns = ['Index', 'Accident_Severity', 'Accident_Date', 'Latitude',
       'Light_Conditions', 'District_Area', 'Longitude',
       'Number_of_Casualties', 'Number_of_Vehicles', 'Road_Surface_Conditions',
       'Road_Type', 'Urban_or_Rural_Area', 'Weather_Conditions',
       'Vehicle_Type']

<h2>Descriptive Analysis</h2>

In [6]:
accidents.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [7]:
accidents.describe().T

count       mean       std        min        25%  \
Latitude              660654.0  52.553866  1.406922  49.914430  51.490690   
Longitude             660653.0  -1.431210  1.383330  -7.516225  -2.332291   
Number_of_Casualties  660679.0   1.357040  0.824847   1.000000   1.000000   
Number_of_Vehicles    660679.0   1.831255  0.715269   1.000000   1.000000   

                            50%        75%        max  
Latitude              52.315641  53.453452  60.757544  
Longitude             -1.411667  -0.232869   1.762010  
Number_of_Casualties   1.000000   1.000000  68.000000  
Number_of_Vehicles     2.000000   2.000000  32.000000

<h2>Checking Null Values</h2>

In [8]:
accidents.isnull()

Index  Accident_Severity  Accident_Date  Latitude  Light_Conditions  \
0       False              False          False     False             False   
1       False              False          False     False             False   
2       False              False          False     False             False   
3       False              False          False     False             False   
4       False              False          False     False             False   
...       ...                ...            ...       ...               ...   
660674  False              False          False     False             False   
660675  False              False          False     False             False   
660676  False              False          False     False             False   
660677  False              False          False     False             False   
660678  False              False          False     False             False   

        District_Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0               False      False                 False               False   
1               False      False                 False               False   
2               False      False                 False               False   
3               False      False                 False               False   
4               False      False                 False               False   
...               ...        ...                   ...                 ...   
660674          False      False                 False               False   
660675          False      False                 False               False   
660676          False      False                 False               False   
660677          False      False                 False               False   
660678          False      False                 False               False   

        Road_Surface_Conditions  Road_Type  Urban_or_Rural_Area  \
0                         False      False                False   
1                         False      False                False   
2                         False       True                False   
3                         False      False                False   
4                         False       True                False   
...                         ...        ...                  ...   
660674                    False      False                False   
660675                    False      False                False   
660676                    False      False                False   
660677                    False      False                False   
660678                    False      False                False   

        Weather_Conditions  Vehicle_Type  
0                    False         False  
1                    False         False  
2                     True         False  
3                    False         False  
4                     True         False  
...                    ...           ...  
660674               False         False  
660675               False         False  
660676               False         False  
660677               False         False  
660678               False         False  

[660679 rows x 14 columns]

In [9]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident_Date                  0
Latitude                      25
Light_Conditions               0
District_Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h2>Filtering a column</h2>

In [10]:
accidents['Accident_Severity']

0         Serious
1         Serious
2         Serious
3         Serious
4          Slight
           ...   
660674     Slight
660675     Slight
660676     Slight
660677    Serious
660678    Serious
Name: Accident_Severity, Length: 660679, dtype: object

In [11]:
accidents['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

In [12]:
accidents[accidents['Accident_Severity'] == 'Slight']

Index Accident_Severity Accident_Date   Latitude  \
4       200701BS66837            Slight      3/9/2019  51.488576   
8       200701BS67515            Slight    31-10-2019  51.493319   
9       200701BS67543            Slight    18-10-2019  51.484539   
13      200701BS67853            Slight     8/11/2019  51.508999   
14      200701BS67859            Slight     1/11/2019  51.485125   
...               ...               ...           ...        ...   
660671  201091NM01053            Slight    29-01-2022  57.572162   
660673  201091NM01190            Slight      2/2/2022  57.585152   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   

              Light_Conditions           District_Area  Longitude  \
4        Darkness - lights lit  Kensington and Chelsea  -0.192487   
8        Darkness - lights lit  Kensington and Chelsea  -0.173572   
9                     Daylight  Kensington and Chelsea  -0.175651   
13       Darkness - lights lit  Kensington and Chelsea  -0.210701   
14       Darkness - lights lit  Kensington and Chelsea  -0.167129   
...                        ...                     ...        ...   
660671                Daylight                Highland  -3.930799   
660673                Daylight                Highland  -3.748312   
660674                Daylight                Highland  -3.467828   
660675  Darkness - no lighting                Highland  -3.809281   
660676                Daylight                Highland  -3.862727   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
4                          1                   2                     Dry   
8                          1                   2                     Dry   
9                          1                   2                     Dry   
13                         3                   2             Wet or damp   
14                         1                   1                     Dry   
...                      ...                 ...                     ...   
660671                     1                   1            Frost or ice   
660673                     2                   2             Wet or damp   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   

                 Road_Type Urban_or_Rural_Area  Weather_Conditions  \
4                      NaN               Urban                 NaN   
8       Single carriageway               Urban  Fine no high winds   
9       Single carriageway               Urban  Fine no high winds   
13      Single carriageway               Urban               Other   
14          One way street               Urban  Fine no high winds   
...                    ...                 ...                 ...   
660671  Single carriageway               Rural               Other   
660673  Single carriageway               Rural  Fine no high winds   
660674  Single carriageway               Rural  Fine no high winds   
660675  Single carriageway               Rural  Fine no high winds   
660676  Single carriageway               Rural  Fine no high winds   

                                 Vehicle_Type  
4                               Other vehicle  
8                                         Car  
9       Motorcycle over 125cc and up to 500cc  
13                                        Car  
14                                        Car  
...                                       ...  
660671    Van / Goods 3.5 tonnes mgw or under  
660673                                    Car  
660674                                    Car  
660675                                    Car  
660676                                    Car  

[563801 rows x 14 columns]

In [13]:
accidents[accidents['Accident_Severity'] != 'Slight']

Index Accident_Severity Accident_Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
5       200701BS67159           Serious    18-09-2019  51.497750   
...               ...               ...           ...        ...   
660629  201091NL00686           Serious     11/2/2022  57.687933   
660645  201091NL05842           Serious    18-10-2022  60.162660   
660672  201091NM01085           Serious    29-01-2022  57.170725   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District_Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
5                          Daylight  Kensington and Chelsea  -0.192556   
...                             ...                     ...        ...   
660629                     Daylight        Shetland Islands  -4.176860   
660645                     Daylight        Shetland Islands  -1.165232   
660672                     Daylight                Highland  -3.852421   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
5                          2                   3                     Dry   
...                      ...                 ...                     ...   
660629                     1                   2                     Dry   
660645                     1                   1                     Dry   
660672                     1                   2                    Snow   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
5       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660629  Single carriageway               Rural      Fine + high winds   
660645  Single carriageway               Rural      Fine + high winds   
660672  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
5                                        Car  
...                                      ...  
660629                                   Car  
660645                                   Car  
660672                                   Car  
660677                 Motorcycle over 500cc  
660

In [14]:
slight_df = accidents[accidents['Accident_Severity'] == 'Slight']
slight_df

Index Accident_Severity Accident_Date   Latitude  \
4       200701BS66837            Slight      3/9/2019  51.488576   
8       200701BS67515            Slight    31-10-2019  51.493319   
9       200701BS67543            Slight    18-10-2019  51.484539   
13      200701BS67853            Slight     8/11/2019  51.508999   
14      200701BS67859            Slight     1/11/2019  51.485125   
...               ...               ...           ...        ...   
660671  201091NM01053            Slight    29-01-2022  57.572162   
660673  201091NM01190            Slight      2/2/2022  57.585152   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   

              Light_Conditions           District_Area  Longitude  \
4        Darkness - lights lit  Kensington and Chelsea  -0.192487   
8        Darkness - lights lit  Kensington and Chelsea  -0.173572   
9                     Daylight  Kensington and Chelsea  -0.175651   
13       Darkness - lights lit  Kensington and Chelsea  -0.210701   
14       Darkness - lights lit  Kensington and Chelsea  -0.167129   
...                        ...                     ...        ...   
660671                Daylight                Highland  -3.930799   
660673                Daylight                Highland  -3.748312   
660674                Daylight                Highland  -3.467828   
660675  Darkness - no lighting                Highland  -3.809281   
660676                Daylight                Highland  -3.862727   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
4                          1                   2                     Dry   
8                          1                   2                     Dry   
9                          1                   2                     Dry   
13                         3                   2             Wet or damp   
14                         1                   1                     Dry   
...                      ...                 ...                     ...   
660671                     1                   1            Frost or ice   
660673                     2                   2             Wet or damp   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   

                 Road_Type Urban_or_Rural_Area  Weather_Conditions  \
4                      NaN               Urban                 NaN   
8       Single carriageway               Urban  Fine no high winds   
9       Single carriageway               Urban  Fine no high winds   
13      Single carriageway               Urban               Other   
14          One way street               Urban  Fine no high winds   
...                    ...                 ...                 ...   
660671  Single carriageway               Rural               Other   
660673  Single carriageway               Rural  Fine no high winds   
660674  Single carriageway               Rural  Fine no high winds   
660675  Single carriageway               Rural  Fine no high winds   
660676  Single carriageway               Rural  Fine no high winds   

                                 Vehicle_Type  
4                               Other vehicle  
8                                         Car  
9       Motorcycle over 125cc and up to 500cc  
13                                        Car  
14                                        Car  
...                                       ...  
660671    Van / Goods 3.5 tonnes mgw or under  
660673                                    Car  
660674                                    Car  
660675                                    Car  
660676                                    Car  

[563801 rows x 14 columns]

In [15]:
serious_df =accidents[accidents['Accident_Severity'] == 'Serious']
serious_df

Index Accident_Severity Accident_Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
5       200701BS67159           Serious    18-09-2019  51.497750   
...               ...               ...           ...        ...   
660629  201091NL00686           Serious     11/2/2022  57.687933   
660645  201091NL05842           Serious    18-10-2022  60.162660   
660672  201091NM01085           Serious    29-01-2022  57.170725   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District_Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
5                          Daylight  Kensington and Chelsea  -0.192556   
...                             ...                     ...        ...   
660629                     Daylight        Shetland Islands  -4.176860   
660645                     Daylight        Shetland Islands  -1.165232   
660672                     Daylight                Highland  -3.852421   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
5                          2                   3                     Dry   
...                      ...                 ...                     ...   
660629                     1                   2                     Dry   
660645                     1                   1                     Dry   
660672                     1                   2                    Snow   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
5       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660629  Single carriageway               Rural      Fine + high winds   
660645  Single carriageway               Rural      Fine + high winds   
660672  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
5                                        Car  
...                                      ...  
660629                                   Car  
660645                                   Car  
660672                                   Car  
660677                 Motorcycle over 500cc  
660

In [16]:
fatal_df = accidents[accidents['Accident_Severity'] == 'Fatal']
fatal_df

Index Accident_Severity Accident_Date   Latitude  \
7       200701BS67370             Fatal     3/10/2019  51.482260   
11      200701BS67747             Fatal     2/11/2019  51.499127   
20      200701BS68301             Fatal    24-10-2019  51.505996   
21      200701BS68302             Fatal    24-11-2019  51.486795   
717     200701CP00032             Fatal    30-01-2019  51.510978   
...               ...               ...           ...        ...   
660470  201091NH00529             Fatal    29-01-2022  58.475726   
660519  201091NJ01533             Fatal      2/2/2022  58.495554   
660565  201091NJ11329             Fatal    15-08-2022  57.871514   
660581  201091NJ14695             Fatal    21-10-2022  58.445475   
660627  201091NL00060             Fatal      6/1/2022  60.241195   

             Light_Conditions           District_Area  Longitude  \
7       Darkness - lights lit  Kensington and Chelsea  -0.179486   
11                   Daylight  Kensington and Chelsea  -0.159652   
20                   Daylight  Kensington and Chelsea  -0.214277   
21                   Daylight  Kensington and Chelsea  -0.170375   
717                  Daylight          City of London  -0.107581   
...                       ...                     ...        ...   
660470               Daylight           Western Isles  -6.278583   
660519               Daylight                Highland  -4.451628   
660565               Daylight                Highland  -4.243081   
660581  Darkness - lights lit                Highland  -3.065535   
660627               Daylight        Shetland Islands  -1.409761   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
7                          3                   2                     Dry   
11                         1                   2                     Dry   
20                         1                   1                     Dry   
21                         5                   1                     Dry   
717                        1                   2             Wet or damp   
...                      ...                 ...                     ...   
660470                     1                   1             Wet or damp   
660519                     1                   1                    Snow   
660565                     1                   1                     Dry   
660581                     1                   1             Wet or damp   
660627                     3                   1                    Snow   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
7       Single carriageway               Urban     Fine no high winds   
11      Single carriageway               Urban     Fine no high winds   
20      Single carriageway               Urban     Fine no high winds   
21      Single carriageway               Urban     Fine no high winds   
717       Dual carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660470  Single carriageway               Rural      Fine + high winds   
660519  Single carriageway               Rural  Snowing no high winds   
660565  Single carriageway               Rural     Fine no high winds   
660581  Single carriageway               Rural     Fine no high winds   
660627  Single carriageway               Rural     Fine no high winds   

                                Vehicle_Type  
7                                        Car  
11      Bus or coach (17 or more pass seats)  
20                                       Car  
21                                       Car  
717                                      Car  
...                                      ...  
660470                                   Car  
660519                                   Car  
660565         Goods 7.5 tonnes mgw and over  
660581                                   Car  
660627   Van / Goods 3.5 tonnes mgw or under  

[8661 rows x 14 columns]

<h2>FILLING UP NULL VALUES WITH STATISTICALLY ACCEPTED VALUES</h2>

In [17]:
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident_Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District_Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [18]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident_Date                  0
Latitude                      25
Light_Conditions               0
District_Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [19]:
accidents['Latitude'] = accidents['Latitude'].fillna(accidents['Latitude'].mean())
accidents['Longitude'] = accidents['Longitude'].fillna(accidents['Longitude'].mean())
accidents['Road_Surface_Conditions'] = accidents['Road_Surface_Conditions'].fillna(accidents['Road_Surface_Conditions'].mode()[0])
accidents['Road_Type'] = accidents['Road_Type'].fillna(accidents['Road_Type'].mode()[0])
accidents['Urban_or_Rural_Area'] = accidents['Urban_or_Rural_Area'].fillna(accidents['Urban_or_Rural_Area'].mode()[0])
accidents['Weather_Conditions'] = accidents['Weather_Conditions'].fillna(accidents['Weather_Conditions'].mode()[0])
accidents.isnull().sum()

Index                      0
Accident_Severity          0
Accident_Date              0
Latitude                   0
Light_Conditions           0
District_Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h2>Changing the Data Type</h2>

In [20]:
accidents['Index'] = accidents['Index'].astype('category')
accidents['Accident_Severity'] = accidents['Accident_Severity'].astype('category')
accidents['Accident_Date'] = accidents['Accident_Date'].astype('category')
accidents['Light_Conditions'] = accidents['Light_Conditions'].astype('category')
accidents['District_Area'] = accidents['District_Area'].astype('category')
accidents['Road_Surface_Conditions'] = accidents['Road_Surface_Conditions'].astype('category')
accidents['Road_Type'] = accidents['Road_Type'].astype('category')
accidents['Urban_or_Rural_Area'] = accidents['Urban_or_Rural_Area'].astype('category')
accidents['Weather_Conditions'] = accidents['Weather_Conditions'].astype('category')
accidents['Vehicle_Type'] = accidents['Vehicle_Type'].astype('category')
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  category
 1   Accident_Severity        660679 non-null  category
 2   Accident_Date            660679 non-null  category
 3   Latitude                 660679 non-null  float64 
 4   Light_Conditions         660679 non-null  category
 5   District_Area            660679 non-null  category
 6   Longitude                660679 non-null  float64 
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  660679 non-null  category
 10  Road_Type                660679 non-null  category
 11  Urban_or_Rural_Area      660679 non-null  category
 12  Weather_Conditions       660679 non-null  category
 13  Vehicle_Type             660679 non-null  ca

<h1>Clearing any Inconsistencies with the Data Sets</h1>

In [21]:
accidents['Accident_Date'] = accidents['Accident_Date'].str.strip()

accidents['Accident_Date'] = accidents['Accident_Date'].astype('str')

accidents['Accident_Date'] = accidents['Accident_Date'].str.replace('/', '-')

accidents['Accident_Date'] = pd.to_datetime(accidents['Accident_Date'], dayfirst = True, errors = 'coerce')

<h1>Adjusting Data Types</h1>

In [22]:
accidents.dtypes

Index                            category
Accident_Severity                category
Accident_Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District_Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

<h1>Extracting date information usinig pandas date time</h1>

In [23]:
accidents['Year'] = accidents['Accident_Date'].dt.year
accidents['Month'] = accidents['Accident_Date'].dt.month
accidents['Day'] = accidents['Accident_Date'].dt.day
accidents['DayOfWeek'] = accidents['Accident_Date'].dt.dayofweek

In [24]:
accidents.isnull().sum()

Index                      0
Accident_Severity          0
Accident_Date              0
Latitude                   0
Light_Conditions           0
District_Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

<h1>Exploratory Data Analytics</h1>
<hr>

<h2>Question 1:</h2>
<h3>How many accidents happened in urban vs rural areas?</h3>

In [25]:
accidents['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h2>Insight:</h2>
<p>Based on the results, urban areas recorded 421,678 accidents while rural areas had 238,990 accidents.</p>
<hr>

<h2>Question 2:</h2>
<h3>What is the most common vehicle type in accidents?</h3>

In [26]:
accidents['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h2>Insight:</h2>
<p>Based on the results, the most common vehicle involved in accidents is cars, with a total of 497,992 cases</p>
<hr>

<h2>Question 3:</h2>
<h3>What is the average number of vehicles involved by road type?</h3>

In [27]:
accidents.groupby('Road_Type')['Number_of_Vehicles'].mean()

Road_Type
Dual carriageway      2.009766
One way street        1.585441
Roundabout            1.901323
Single carriageway    1.794803
Slip road             1.917483
Name: Number_of_Vehicles, dtype: float64

<h2>Insight:</h2>
<p>Based on the results, major roads and highways tend to involve more vehicles on average compared to smaller roads.</p>
<hr>

<h2>Question 4:</h2>
<h3>How many accidents happened under each light condition?</h3>

In [28]:
accidents['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h2>Insight:</h2>
<p>Based on the results, most accidents occurred during daylight while fewer accidents happened under poor lighting or darkness</p>
<hr>

<h2>Question 5:</h2>
<h3>What is the average number of casualties per accident severity level?</h3>

In [29]:
accidents.groupby('Accident_Severity')['Number_of_Casualties'].mean()

Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h2>Insight:</h2>
<p>Based on the results, accidents with higher severity levels result in a greater number of casualties.</p>
<hr>

<h2>Question 6:</h2>
<h3>What is the correlation between the number of vehicles and the number of casualties?</h3>

In [30]:
vehicles_casualties = (accidents['Number_of_Vehicles'].corr(accidents['Number_of_Casualties']))
vehicles_casualties

np.float64(0.22888886126927557)

<h2>Insight:</h2>
<p>The results show a weak positive relationship, meaning more vehicles can lead to slightly more casualties</p>
<hr>

<h2>Question 7:</h2>
<h3>Which Accident Severity has the highest average number of casualties?</h3>

In [31]:
slight_df['Number_of_Casualties'].mean()

np.float64(1.3314023919787301)

In [32]:
serious_df['Number_of_Casualties'].mean()

np.float64(1.46727954929322)

In [33]:
fatal_df['Number_of_Casualties'].mean()

np.float64(1.9031289689412307)

<h2>Insight:</h2>
<p>The resutls show that fatal accidents have the highest average casualties, followed by serious and slight accidents.</p>
<hr>

<h2>Question 8:</h2>
<h3>Which severity level involves the highest average number of vehicles?</h3>

In [34]:
slight_df['Number_of_Vehicles'].mean()

np.float64(1.8558640371336694)

In [35]:
serious_df['Number_of_Vehicles'].mean()

np.float64(1.6783273065282203)

In [36]:
fatal_df['Number_of_Vehicles'].mean()

np.float64(1.7869760997575337)

<h2>Insight: </h2>
<p> The results show that serious accidents tend to involve slightly more vehicles compared to slight or fatal ones.</p>
<hr>

<h2>Question 9: </h2>
<h3>Which weather condition has the highest total number of accidents?</h3>

In [37]:
accidents['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h2>Insight:</h2>
<p>The results show that the most common weather condition highlights typical driving environments for accidents.</p>
<hr>

<h2>Question 10</h2>
<h3>How many accidents happened in each month?</h3>

In [38]:
accidents.groupby('Month')['Index'].count()

Month
1     52872
2     49491
3     54086
4     51744
5     56352
6     56481
7     57445
8     53913
9     56455
10    59580
11    60424
12    51836
Name: Index, dtype: int64

<h2>Insight:</h2>
<p>In the results months like August or July may show higher accidents due to summer travel, while winter months like February have fewer due to less road activity.</p>
<hr>

<h2>Question 11: </h2>
<h3>What is the average number of vehicles in fatal accidents by road type?</h3>

In [39]:
fatal_df.groupby('Road_Type')['Number_of_Vehicles'].mean()

Road_Type
Dual carriageway      1.932231
One way street        1.357895
Roundabout            1.549296
Single carriageway    1.761759
Slip road             1.571429
Name: Number_of_Vehicles, dtype: float64

<h2>Insight:</h2>
<p>The results show that dual carriageways in fatal accidents average more vehicles than single carriageways , likely due to higher speeds and traffic density.</p>
<hr>

<h2>Question 12: </h2>
<h3>How many accidents involved buses or coaches?</h3>

In [41]:
accidents[accidents['Vehicle_Type'] == 'Bus or coach (17 or more pass seats)']['Index'].count()

np.int64(25878)

<h2>Insight:</h2>
<p>In the results buses/coaches are involved in 25,879 accidents, a small but significant portion due to their capacity for multiple casualties.</p>
<hr>

<h2>Question 13: </h2>
<h3>What is the total number of casualties in accidents during darkness?</h3>

In [42]:
accidents[accidents['Light_Conditions'] != 'Daylight']['Number_of_Casualties'].sum()

np.int64(246679)

<h2>Insight: </h2>
<p>In the results darkness related accidents contribute 246,679 casualties, emphasizing the need for better lighting or driver caution at night.</p>
<hr>

<h2>Question 14:</h2>
<h3>What is the total number of accidents per year?</h3>

In [45]:
accidents.groupby('Year')['Index'].count()

Year
2019    182115
2020    170591
2021    163554
2022    144419
Name: Index, dtype: int64

<h2>Insight:</h2>
<p>The Results show that accidents likely peaked around 2019, with potential dips in 2020-2022 due to reduced travel during COVID-19 restrictions</p>
<hr>

<h2>Question 15:</h2>
<h3>What is the total number of casualties in urban vs. rural areas?</h3>

In [46]:
accidents.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].sum()

Urban_or_Rural_Area
Rural          353515
Unallocated        13
Urban          543040
Name: Number_of_Casualties, dtype: int64

<h2>Insight:</h2>
<p>The results show that urban areas have more total casualties than rural due to higher accident frequency, despite rural crashes often being more severe.</p>
<hr

<h2>Question 16:</h2>
<h3>What is the total number of accidents by vehicle type and accident severity?</h3>

In [47]:
accidents.groupby(['Vehicle_Type', 'Accident_Severity'])['Index'].count()

Vehicle_Type                           Accident_Severity
Agricultural vehicle                   Fatal                    21
                                       Serious                 282
                                       Slight                 1644
Bus or coach (17 or more pass seats)   Fatal                   325
                                       Serious                3373
                                       Slight                22180
Car                                    Fatal                  6577
                                       Serious               66461
                                       Slight               424954
Data missing or out of range           Fatal                     0
                                       Serious                   0
                                       Slight                    6
Goods 7.5 tonnes mgw and over          Fatal                   216
                                       Serious                2321
     

<h2>Insight:</h2>
<p>It shows that cars dominate across all severities accidents, but motorcycles show a higher proportion in serious and fatal cases, aligning with stats where motorcyclists account for 19% of fatalities.</p>
<hr> 

<h2>Question 17:</h2>
<h3>What is the total number of accidents by year and accident severity?</h3>

In [49]:
accidents.groupby(['Year', 'Accident_Severity'])['Index'].count()

Year  Accident_Severity
2019  Fatal                  2714
      Serious               24322
      Slight               155079
2020  Fatal                  2341
      Serious               23121
      Slight               145129
2021  Fatal                  2057
      Serious               21997
      Slight               139500
2022  Fatal                  1549
      Serious               18777
      Slight               124093
Name: Index, dtype: int64

<h2>Insight:</h2>
<p>It shows that in 2022 slight accidents were 124,093, serious 18,777, and fatal 1,549, showing a potential decline in fatals compared to 2019, possibly from safety improvements.</p>
<hr>

<h2>Question 18:</h2>
<h3>What is the average number of vehicles in accidents during daylight vs. darkness?</h3>

In [53]:
accidents.groupby('Light_Conditions')['Number_of_Vehicles'].mean()

Light_Conditions
Darkness - lighting unknown    1.758482
Darkness - lights lit          1.775498
Darkness - lights unlit        1.752654
Darkness - no lighting         1.599781
Daylight                       1.865385
Name: Number_of_Vehicles, dtype: float64

<h2>Insight:</h2>
<p>It shows that dayylight accidents average slightly more vehicles than darkness, possibly due to higher traffic volumes during the day.</p>
<hr>

<h2>Question 19:</h2>
<h3>How many serious accidents occurred in each light condition?</h3>

In [56]:
serious_df.groupby('Light_Conditions')['Index'].count()

Light_Conditions
Darkness - lighting unknown      794
Darkness - lights lit          19130
Darkness - lights unlit          360
Darkness - no lighting          7174
Daylight                       60759
Name: Index, dtype: int64

<h2>Insight:</h2>
<p>The results show that daylight dominates serious accidents, but darkness with no lighting has a notable share, indicating visibility as a risk factor.</p>
<hr>

<h2>Question 20:</h2>
<h3>What is the average number of vehicles in accidents during fine weather?</h3>

In [57]:
accidents[accidents['Weather_Conditions'] == 'Fine no high winds']['Number_of_Vehicles'].mean()

np.float64(1.8359067910499371)

<h2>Insight:</h2>
<p> The results show that fine weather accidents average 1.8 vehicles, similar to overall trends, suggesting traffic volume rather than weather drives multi-vehicle incidents.</p>
<hr>